### PR TITLE
Add TensorBoard analysis script to Universal Checkpointing Example

### PR DESCRIPTION
**Pull request by lekurile**
This PR adds a tensorboard analysis script to the Universal Checkpointing example. The script recursively searches for all tensorboard output logs, creates corresponding .csv output files of the `lm loss` of each run, and creates a plot of all the different runs.

Example output:
image

image

### CODE REVIEW

**Review by loadams — APPROVED** [review]

File: examples_deepspeed/universal_checkpointing/tb_analysis/tb_analysis_script.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+from utils import get_analyzer, find_files_recursive
+from arguments import parser
+
+args = parser.parse_args()
+
+if args.use_sns:
+    try:
+        import seaborn as sns
+        sns.set()
+    except ImportError as e:
```
Comment (mrwyattii) [inline examples_deepspeed/universal_checkpointing/tb_analysis/tb_analysis_script.py:20]: If a user provides `--use_sns` I think we should error out with the package is not found.

File: examples_deepspeed/universal_checkpointing/tb_analysis/tb_analysis_script.py
Comment (mrwyattii) [inline examples_deepspeed/universal_checkpointing/tb_analysis/tb_analysis_script.py:1]: We should place the main content of this script in `def main():` and then add:
```python
if __name__ == "__main__":
    main()
```

File: examples_deepspeed/universal_checkpointing/tb_analysis/tb_analysis_script.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        print(df)
+
+if not args.skip_plot:
+    plt.legend()
+    plt.title(args.plot_title)
+    plt.xlabel(args.plot_x_label)
+    plt.ylabel(args.plot_y_label)
+    plt.savefig(args.plot_name)
+
+print(tb_log_paths)
```
Comment (mrwyattii) [inline examples_deepspeed/universal_checkpointing/tb_analysis/tb_analysis_script.py:54]: Is this a leftover debug print?

File: examples_deepspeed/universal_checkpointing/tb_analysis/tb_analysis_script.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+    x = [x.step for x in events]
+    y = [x.value for x in events]
+
+    plt.plot(x, y, label=f'{analyzer.get_label_name()}')
+
+    if not args.skip_csv:
+        df = pd.DataFrame({"step": x, "value": y})
+        df.to_csv(f"{args.csv_name}{analyzer.get_csv_filename()}.csv")
+        print(df)
```
Comment (mrwyattii) [inline examples_deepspeed/universal_checkpointing/tb_analysis/tb_analysis_script.py:45]: debug print?

File: examples_deepspeed/universal_checkpointing/tb_analysis/uc_analysis.py [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+    def __init__(self):
+        self._name = "universal_checkpointing"
+
+    def set_names(self, path_name):
+        match = re.match(self.path_regex(), path_name)
+        tp = match.group(1)
+        pp = match.group(2)
+        dp = match.group(3)
+        sp = match.group(4)
```
Comment (mrwyattii) [inline examples_deepspeed/universal_checkpointing/tb_analysis/uc_analysis.py:20]:
```suggestion
        match = re.match(self.path_regex(), path_name)
        if not match:
            raise ValueError(f"Path ({path_name}) did not match regex ({self.path_regex()})")
        tp, pp, dp, sp = match.groups()
```

File: examples_deepspeed/universal_checkpointing/tb_analysis/utils.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+
+    Args:
+        directory (str): The path to the directory to search.
+        file_affix (str): The desired file affix
+
+    Returns:
+        list: A list of paths to matching files.
+    """
+    matching_paths = []
+    for root, _, files in os.walk(directory):
```
Comment (mrwyattii) [inline examples_deepspeed/universal_checkpointing/tb_analysis/utils.py:22]: The function name indicates that it is recursive. Do we mean to drop the directory list and not search sub directories?

File: examples_deepspeed/universal_checkpointing/tb_analysis/tb_analysis_script.py (hunk 1 above)
Comment (lekurile) [reply examples_deepspeed/universal_checkpointing/tb_analysis/tb_analysis_script.py]: Addressed, thanks.

File: examples_deepspeed/universal_checkpointing/tb_analysis/tb_analysis_script.py
Comment (lekurile) [reply examples_deepspeed/universal_checkpointing/tb_analysis/tb_analysis_script.py:1]: Updated.

File: examples_deepspeed/universal_checkpointing/tb_analysis/tb_analysis_script.py (hunk 2 above)
Comment (lekurile) [reply examples_deepspeed/universal_checkpointing/tb_analysis/tb_analysis_script.py]: Removed.

File: examples_deepspeed/universal_checkpointing/tb_analysis/tb_analysis_script.py (hunk 3 above)
Comment (lekurile) [reply examples_deepspeed/universal_checkpointing/tb_analysis/tb_analysis_script.py]: Removed.

File: examples_deepspeed/universal_checkpointing/tb_analysis/uc_analysis.py (hunk 4 above)
Comment (lekurile) [reply examples_deepspeed/universal_checkpointing/tb_analysis/uc_analysis.py]: Updated, thanks.

File: examples_deepspeed/universal_checkpointing/tb_analysis/utils.py (hunk 5 above)
Comment (lekurile) [reply examples_deepspeed/universal_checkpointing/tb_analysis/utils.py:22]: Updated the function name and description to remove the notion of recursion of the function itself and instead just reference that `os.walk()` is used to search for files.

**Review by tjruwase — APPROVED** [review]

**Pull request merged**